File: utils/py/networking.py
'''
networking.py - The Python networking engine for communication between quad's client and server modules.

    Copyright (c) 2016 Dariush Hasanpour

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
'''

import time
import json
import struct
import socket_server as server

# define timeout second
TIMEOUT_SEC      = 1.0

def sendFloats(client, data):
    client.send(struct.pack('%sf' % len(data), *data))

def unpackFloats(msg, nfloats):
    return struct.unpack('f'*nfloats, msg)

def rcvFloats(client, nfloats):
    # We use 32-bit floats
    msgsize = 4 * nfloats

    # Implement timeout
    start_sec = time.time()
    remaining = msgsize
    msg = ''
    while remaining > 0:
        msg += client.recv(remaining)
        remaining -= len(msg)
        if (time.time()-start_sec) > TIMEOUT_SEC:
            return None

    return unpackFloats(msg, nfloats)

def sendString(client, _str):
	sendFloats(client, [len(_str)]);
	client.send(_str)

def rcvString(client):
    return client.recv(int(rcvFloats(client, 1)[0]))

def sendJSON(client, data):
	sendString(client, json.dumps(data))

def rcvJSON(client):
	return json.loads(rcvString(client))

def initiate_handshake(client):
    if(rcvString(client) != "220 EHLO SERVER"):
        print("Server: Wrong Handshake!");
        exit(-1);
    else:
        sendString(client, "250 Acknowledged Handshake")
    # handshake went OK, rcv metadata
    meta = rcvJSON(client);
    print "Server: valid handshake."
    return meta 

File: utils/lua/networking.lua
--[[
networking.lua - The Lua networking engine for communication between quad's client and server modules.

    Copyright (c) 2016 Dariush Hasanpour

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
]]--

JSON = (loadfile(PWD..'/utils/lua/JSON.lua'))()

function server_connect(SERVER_EXECUTABLE)
    local portNb = simGetInt32Parameter(sim_intparam_server_port_next)
    local portStart = simGetInt32Parameter(sim_intparam_server_port_start)
    local portRange = simGetInt32Parameter(sim_intparam_server_port_range)
    local newPortNb = portNb + 1
    if (newPortNb >= portStart + portRange) then
        newPortNb = portStart
    end

    simSetInt32Parameter(sim_intparam_server_port_next, newPortNb)

    local serverResult = simLaunchExecutable(SERVER_EXECUTABLE, portNb, 0) 

    -- Attempt to launch the executable server script
    if (serverResult==-1) then
        simDisplayDialog('Error',
            'Server '..SERVER_EXECUTABLE..' could not be launched. &&n'..
            'Please make sure that it exists and is executable. Then stop and restart the simulation.',
            sim_dlgstyle_message, false,  nil, {0.8,0,0,0,0,0}, {0.5,0,0,1,1,1})
        simPauseSimulation()
    end

    -- On success, attempt to connect to server
    while (serverResult ~= -1 and simGetSimulationState() ~= sim_simulation_advancing_abouttostop) do
        -- The executable could be launched. Now build a socket and connect to the server:
        local socket = require("socket")
        local server = socket.tcp()
        if server:connect('127.0.0.1', portNb) == 1 then
            return server
        end
    end
end

function server_close(server)
    server:close()
end

function initiate_handshake(server)
    sendString(server, "220 EHLO SERVER")

    if not string.match(rcvString(server), "250 .+") then
        print("Client: Wrong Handshake!")
        simStopSimulation()
    else
        print("Client: valid handshake.")
    end

    -- handshake went OK, send metadata
    sendJSON(server, { ["PWD"] = PWD, ["VERSION"] = VERSION })
end

function sendFloats(server, data)
    server:send(simPackFloats(data))
end

function rcvFloats(server, count)
    data = server:receive(4 * count)
    return simUnpackFloats(data)
end

function sendString(server, str)
    sendFloats(server, {string.len(str)})
    server:send(str)
end

function rcvString(server)
    return server:receive(rcvFloats(server, 1)[1]);
end

function sendJSON(server, data)
    sendString(server, JSON:encode(data))
end

function rcvJSON(server)
    return JSON:decode(rcvString(server))
end